examples/codenames/Codenames_RL.ipynb

```diff
@@ -0,0 +1,12542 @@
+      },
+      "outputs": [],
+      "source": [
+        "# ────────────────────  5. RL / reward helpers  ────────────────────────────\n",
+        "def create_mc(player: LLMSpymaster | LLMOperative) -> List[dict]:\n",
+        "    \"\"\"\n",
+        "    Merge messages & choices into a single list for art.\n",
+        "    \"\"\"\n",
+        "    mc = []\n",
+        "    choice_i = 0\n",
+        "    for msg in player.conversation:\n",
+        "        if msg[\"role\"] == \"assistant\":\n",
+        "            mc.append(player.conversation_choices[choice_i])\n",
+        "            choice_i += 1\n",
+        "        else:\n",
+        "            mc.append(msg)\n",
+        "    return mc\n",
+        "\n",
+        "\n",
+        "def create_margin_reward(\n",
+        "    player: LLMSpymaster | LLMOperative,\n",
+        "    role_key: str,\n",
+        "    gs: GameState,\n",
+        "    winning_color: str,\n",
+        "    reason: str,\n",
+        ") -> float:\n",
+        "    num_start_words = 9 if player.team == \"blue\" else 8\n",
+        "    reward = 0.0\n",
+        "\n",
+        "    if gs.game_ended_in_turn_limit:\n",
+        "        reward += MARGIN_REWARD_DICT[\"turn_limit\"]\n",
+        "        return reward\n",
+        "    if gs.game_ended_in_error and role_key == f\"{gs.game_ended_in_error['team']}_{gs.game_ended_in_error['role']}\":\n",
+        "        reward += MARGIN_REWARD_DICT[\"error\"]\n",
+        "\n",
+        "    if gs.game_ended_in_assassin and player.team != winning_color:\n",
+        "        reward += MARGIN_REWARD_DICT[\"assassin\"]\n",
+        "\n",
+        "    # proportion of own words guessed\n",
+        "    reward += (\n",
+        "        (num_start_words - gs.words_left[player.team])\n",
+        "        / num_start_words\n",
+        "        * MARGIN_REWARD_DICT[\"num_words_multiplier\"]\n",
+        "    )\n",
+        "\n",
+        "    # margin\n",
+        "    loser = \"red\" if winning_color == \"blue\" else \"blue\"\n",
+        "    margin = gs.words_left[loser]\n",
+        "    reward += margin * MARGIN_REWARD_DICT[\"margin_multiplier\"] * (\n",
+        "        1 if player.team == winning_color else -1\n",
+        "    )\n",
+        "    return reward"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "u2v5YiKPwu9b"
+      },
+      "source": [
+        "# Defining a rollout\n",
+        "<a name=\"Rollout\"></a>\n",
+        "\n",
+        "In each training iteration, we start with 10 different Codenames boards. For every board, the model plays 20 complete self‑play games (rollouts). Each rollout produces four separate records of experience—one for every role in the game—so a single board yields 80 trajectories in total. We then gather those 80 trajectories into 4 role‑based groups per board. Across the 10 boards, this gives us 40 trajectory groups that the learning algorithm uses in that iteration."
```

Comment:
This is super interesting, but could you spend a little more time breaking up the concepts clearly? I realize this isn't perfect, but seeing a math equation would make it easier for me to follow how you arrived at your total number of trajectory groups.

* `boards` = 10
* `games_per_board` = 20
* `roles` = 4
  * (red spymaster, red operative, blue spymaster, blue operative)

`total_trajectories` = `boards` * `games_per_board` * `roles` = 800
`trajectory_groups` = `boards` * `roles` = 40